pkg/occlient/projects.go

```diff
@@ -0,0 +1,236 @@
+							break
+						}
+					}
+
+				} else {
+					watchErrorChannel <- errors.New("unable to convert event object to Project")
+					break
+				}
+
+			}
+			close(projectChannel)
+			close(watchErrorChannel)
+		}()
+
+		select {
+		case val := <-projectChannel:
+			klog.V(3).Infof("Deletion information for project: %+v", val)
+			return nil
+		case err := <-watchErrorChannel:
+			return err
+		case <-time.After(waitForProjectDeletionTimeOut):
+			return errors.Errorf("waited %s but couldn't delete project %s in time", waitForProjectDeletionTimeOut, name)
+		}
+
+	}
+
+	// Return nil since we don't bother checking for the watcher..
+	return nil
+}
+
+// CreateNewProject creates project with given projectName
+func (c *Client) CreateNewProject(projectName string, wait bool) error {
+	// Instantiate watcher before requesting new project
+	// If watched is created after the project it can lead to situation when the project is created before the watcher.
+	// When this happens, it gets stuck waiting for event that already happened.
+	var watcher watch.Interface
+	var err error
+	if wait {
+		watcher, err = c.projectClient.Projects().Watch(metav1.ListOptions{
+			FieldSelector: fields.Set{"metadata.name": projectName}.AsSelector().String(),
+		})
+		if err != nil {
+			return errors.Wrapf(err, "unable to watch new project %s creation", projectName)
+		}
+		defer watcher.Stop()
+	}
+
+	projectRequest := &projectv1.ProjectRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: projectName,
+		},
+	}
+	_, err = c.projectClient.ProjectRequests().Create(projectRequest)
+	if err != nil {
+		return errors.Wrapf(err, "unable to create new project %s", projectName)
+	}
+
+	if watcher != nil {
+		for {
+			val, ok := <-watcher.ResultChan()
+			if !ok {
+				break
+			}
+			if prj, ok := val.Object.(*projectv1.Project); ok {
+				klog.V(3).Infof("Status of creation of project %s is %s", prj.Name, prj.Status.Phase)
+				switch prj.Status.Phase {
+				//prj.Status.Phase can only be "Terminating" or "Active" or ""
+				case corev1.NamespaceActive:
+					if val.Type == watch.Added {
+						klog.V(3).Infof("Project %s now exists", prj.Name)
+						return nil
+					}
+					if val.Type == watch.Error {
+						return fmt.Errorf("failed watching the deletion of project %s", prj.Name)
+					}
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// IsProjectSupported checks if Project resource type is present on the cluster
+func (c *Client) IsProjectSupported() (bool, error) {
+	return c.isResourceSupported("project.openshift.io", "v1", "projects")
+}
+
+// GetCurrentProjectName returns the current project name
+func (c *Client) GetCurrentProjectName() string {
+	return c.Namespace
+}
+
+// SetCurrentProject sets the given projectName to current project
+func (c *Client) SetCurrentProject(projectName string) error {
+	rawConfig, err := c.KubeConfig.RawConfig()
+	if err != nil {
+		return errors.Wrapf(err, "unable to switch to %s project", projectName)
```

Comment:
We're not really trying to switch to a project specified by `projectName` in this step. It's kinda misleading to report so at this step, no? Can we change the error message to something that's more relevant for this step?

Reply:
Fixed and returning the complete error instead now.